modules/checklist/steps-manager.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace Elementor\Modules\Checklist;
+
+use Elementor\Modules\Checklist\Steps\Create_Pages;
+use Elementor\Modules\Checklist\Steps\Step_Base;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Steps_Manager {
+	/** @var Step_Base[] $step_instances */
+	private array $step_instances = [];
+
+	private Module $module;
+
+	public function __construct( Module $module ) {
+		$this->module = $module;
+		$this->register_steps();
+	}
+
+	/**
+	 * Gets formatted and ordered array of step ( step data, is_marked_completed and is_completed )
+	 *
+	 * @return array
+	 */
+	public function get_steps_for_frontend() : array {
+		$formatted_steps = [];
+
+		foreach ( $this->step_instances as $step ) {
+			$formatted_steps[] = $step->get_step_config_for_frontend();
+		}
+
+		return $formatted_steps;
+	}
+
+	/**
+	 * Marks a step as completed, returns true if the step was found and marked or false otherwise
+	 *
+	 * @param string $step_id
+	 *
+	 * @return void
+	 */
+	public function mark_step_as_completed( string $step_id ) : void {
+		foreach ( $this->step_instances as $step ) {
+			if ( $step->get_id() === $step_id ) {
+				$step->mark_as_completed();
+
+				return;
+			}
+		}
+	}
+
+	/**
+	 * Unmarks a step as completed, returns true if the step was found and unmarked or false otherwise
+	 *
+	 * @param string $step_id
+	 *
+	 * @return void
+	 */
+	public function unmark_step_as_completed( string $step_id ) : void {
+		foreach ( $this->step_instances as $step ) {
+			if ( $step->get_id() === $step_id ) {
+				$step->unmark_as_completed();
+
+				return;
+			}
+		}
+	}
+
+	/**
+	 * Maybe marks a step as completed (depending on if source allows it), returns true if the step was found and marked or false otherwise
+	 *
+	 * @param $step_id
+	 *
+	 * @return void
+	 */
+	public function maybe_set_step_as_immutable_completed( string $step_id ) : void {
+		foreach ( $this->step_instances as $step ) {
+			if ( $step->get_id() === $step_id ) {
+				$step->maybe_mark_as_completed();
+
+				return;
+			}
+		}
+	}
+
+	public function get_step_by_id( string $step_id ) : ?Step_Base {
+		foreach ( $this->step_instances as $step ) {
+			if ( $step->get_id() === $step_id ) {
+				return $step;
+			}
+		}
```

Comment:
Do you need a loop here? Just check if `$this->step_instances[ $step_id ]` is set, and then return it?

Reply:
The instances are not associative, I thought we should have them ordered but if you prefer we can make it associative

Reply:
I'll switch it to be associative array

Reply:
Done